src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -0,0 +1,142 @@
+									  Query *selectQuery, EState *executorState);
+static void ExecuteQuery(Query *query, ParamListInfo params, DestReceiver *dest);
+
+
+/*
+ * CoordinatorInsertSelectExecScan executes an INSERT INTO distributed_table
+ * SELECT .. query by setting up a DestReceiver that copies tuples into the
+ * distributed table and then executing the SELECT query using that DestReceiver
+ * as the tuple destination.
+ */
+TupleTableSlot *
+CoordinatorInsertSelectExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
+	{
+		EState *executorState = scanState->customScanState.ss.ps.state;
+		MultiPlan *multiPlan = scanState->multiPlan;
+		Query *selectQuery = multiPlan->insertSelectQuery;
+		List *insertTargetList = multiPlan->insertTargetList;
+		Oid targetRelationId = multiPlan->targetRelationId;
+
+		ereport(DEBUG1, (errmsg("Collecting INSERT ... SELECT results locally")));
+
+		ExecuteSelectIntoRelation(targetRelationId, insertTargetList, selectQuery,
+								  executorState);
+
+		scanState->finishedRemoteScan = true;
+	}
+
+	resultSlot = ReturnTupleFromTuplestore(scanState);
+
+	return resultSlot;
+}
+
+
+/*
+ * ExecuteSelectIntoRelation executes given SELECT query and inserts the
+ * results into the target relation, which is assumed to be a distributed
+ * table.
+ */
+static void
+ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+						  Query *selectQuery, EState *executorState)
+{
+	ParamListInfo paramListInfo = executorState->es_param_list_info;
+
+	ListCell *insertTargetCell = NULL;
+	List *columnNameList = NIL;
+	bool stopOnFailure = false;
+
+	CitusCopyDestReceiver *copyDest = NULL;
+
+	BeginOrContinueCoordinatedTransaction();
+
+	/* build the list of column names for the COPY statement */
+	foreach(insertTargetCell, insertTargetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+
+		columnNameList = lappend(columnNameList, insertTargetEntry->resname);
+	}
+
+	/* set up a DestReceiver that copies into the distributed table */
+	copyDest = CreateCitusCopyDestReceiver(targetRelationId, columnNameList,
+										   executorState, stopOnFailure);
+
+	ExecuteQuery(selectQuery, paramListInfo, (DestReceiver *) copyDest);
+
+	executorState->es_processed = copyDest->tuplesSent;
+}
+
+
+/*
+ * ExecuteQuery plans and executes a query and sends results to the given
+ * DestReceiver.
+ */
+static void
+ExecuteQuery(Query *query, ParamListInfo params, DestReceiver *dest)
+{
+	PlannedStmt *queryPlan = NULL;
+	Portal portal = NULL;
+	int eflags = 0;
+	long count = FETCH_ALL;
+
+	/* create a new portal for executing the query */
+	portal = CreateNewPortal();
+
+	/* don't display the portal in pg_cursors, it is for internal use only */
+	portal->visible = false;
+
+	/* plan the subquery, this may be another distributed query */
+	queryPlan = pg_plan_query(query, 0, params);
```

Comment:
> What about passing CURSOR_OPT_PARALLEL_OK to pg_plan_query? That seems similar to this discussion?

Good idea!

> Also, shall we obey the citus.binary_master_copy_format flag here?

That should happen automatically if selecting from a distributed table.